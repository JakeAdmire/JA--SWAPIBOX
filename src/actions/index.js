export const addFavorite = (film) => ({
  type: 'ADD_FAVORITE',
  film
})

export const removeFavorite = (film) => ({
  type: 'REMOVE_FAVORITE',
  film
})

export const setCategory = (event) => ({
  type: 'SET_CATEGORY',
  event
})