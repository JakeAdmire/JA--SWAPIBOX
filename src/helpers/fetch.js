export const fetchHappens = async (url) => {
  try {
    const results = await fetch(url);
    const data = await results.json();
    return data.results;
  } catch(error) {
    return error;
  }
}