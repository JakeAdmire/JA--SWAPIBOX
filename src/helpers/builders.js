import React from 'react';
import { fetchHappens } from './fetch';

export const peopleGatherer = async (person) => {
  const homeworld = await fetchHappens(person.homeworld);
  const species = await fetchHappens(person.species);

  return {
    name: person.name, 
    homeworld: homeworld.name,
    species: species.name,
    language: species.language,
    population: homeworld.population,
    category: 'people'
  };
}

export const peopleCreator = (info) => {
  console.log(info);
  const { name, homeworld, species, population } = info;
  return (
    <div className="person card">
      <h3>Name: {name}</h3>
      <p>Species: {species}</p>
      <p>Homeworld: {homeworld}</p>
      <p>Planet population: {population}</p>
    </div>
  )
}

export const planetsGatherer = (planet) => {
  return {
    name: planet.name, 
    terrain: planet.terrain,
    population: planet.population,
    climate: planet.climate,
    residents: planet.residents.length,
    category: 'planets'
  };
}

export const planetsCreator = (info) => {
  console.log(info);
  const { name, terrain, population, climate, residents } = info;
  return (
    <div className="planet card">
      <h3>Name: {name}</h3>
      <p>Terrain: {terrain}</p>
      <p>Population: {population}</p>
      <p>Climate: {climate}</p>
      <p>Residents: {residents}</p>
    </div>
  )
}