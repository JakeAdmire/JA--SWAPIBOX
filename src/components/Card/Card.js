import React from 'react';
import { connect } from 'react-redux';
// import PropTypes from 'prop-types';
import { addFavorite, removeFavorite } from '../../actions';
import * as builders from '../../helpers/builders';

export function Card(props) {

  let card = builders[`${props.category}Creator`](props);
  let fave = 'fave-button';

  if (props.faves.includes(props.name)) {
    console.log('active');
    fave = 'fave-button active';
  }

  const toggleFavorite = () => {
    if (props.faves.includes(props.name)) {
      props.removeFavorite(props.name)
    } else {
      props.addFavorite(props.name);
    }
    console.log('faves props:', props.faves);
  }

  return (
    <div>
      <button onClick={toggleFavorite} className={fave}>FAVE</button>
      { card }
    </div>
  )
}

// Card.propTypes = {
//   faves: PropTypes.array.isRequired,
//   addFavorite: PropTypes.func.isRequired,
//   removeFavorite: PropTypes.func.isRequired,
//   category: PropTypes.string.isRequired
// };

export const mapStateToProps = (state) => ({
  faves: state.faves
})

export const mapDispatchToProps = (dispatch) => ({
  addFavorite: (film) => dispatch(addFavorite(film)),
  removeFavorite: (film) => dispatch(removeFavorite(film))
})

export default connect(mapStateToProps, mapDispatchToProps)(Card);