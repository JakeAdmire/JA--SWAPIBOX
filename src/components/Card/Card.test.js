import { enzyme, shallow } from 'enzyme';
import React from 'react';

import { Card, mapStateToProps, mapDispatchToProps } from './Card';
import { addFavorite, removeFavorite } from '../../actions';
import * as builders from '../../helpers/builders';

describe('Card', () => {

  let wrapper;

  let addFavorite = jest.fn();
  let removeFavorite = jest.fn();

  let mockProps = {
    faves: [],
    addFavorite,
    removeFavorite,
    category: 'people'
  }

  wrapper = shallow(
    <Card {...mockProps} />
  )

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  }) 

})

describe('mapStateToProps', () => {

  it('should return a faves array and as props', () => {
    const mockState = {
      faves: []
    }

    const results = mapStateToProps(mockState)
    expect(results).toEqual(mockState)
  })

})

describe('mapDispatchToProps', () => {

  const mockFilm = {};
  const mockDispatch = jest.fn();

  it('should dispatch the addFavorite method', () => {
    const actionToDispatch = addFavorite(mockFilm);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.addFavorite(mockFilm);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  })

  it('should dispatch the removeFavorite method', () => {
    const actionToDispatch = removeFavorite(mockFilm);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.removeFavorite(mockFilm);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  })

})