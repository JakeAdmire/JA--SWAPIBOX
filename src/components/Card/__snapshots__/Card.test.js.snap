// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Card
    addFavorite={[MockFunction]}
    category="people"
    faves={Array []}
    removeFavorite={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <button
          className="fave-button"
          onClick={[Function]}
        >
          FAVE
        </button>,
        <div
          className="person card"
        >
          <h3>
            Name: 
          </h3>
          <p>
            Species: 
          </p>
          <p>
            Homeworld: 
          </p>
          <p>
            Planet population: 
          </p>
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "FAVE",
          "className": "fave-button",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "FAVE",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h3>
              Name: 
            </h3>,
            <p>
              Species: 
            </p>,
            <p>
              Homeworld: 
            </p>,
            <p>
              Planet population: 
            </p>,
          ],
          "className": "person card",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Name: ",
                undefined,
              ],
            },
            "ref": null,
            "rendered": Array [
              "Name: ",
              undefined,
            ],
            "type": "h3",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Species: ",
                undefined,
              ],
            },
            "ref": null,
            "rendered": Array [
              "Species: ",
              undefined,
            ],
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Homeworld: ",
                undefined,
              ],
            },
            "ref": null,
            "rendered": Array [
              "Homeworld: ",
              undefined,
            ],
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Planet population: ",
                undefined,
              ],
            },
            "ref": null,
            "rendered": Array [
              "Planet population: ",
              undefined,
            ],
            "type": "p",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <button
            className="fave-button"
            onClick={[Function]}
          >
            FAVE
          </button>,
          <div
            className="person card"
          >
            <h3>
              Name: 
            </h3>
            <p>
              Species: 
            </p>
            <p>
              Homeworld: 
            </p>
            <p>
              Planet population: 
            </p>
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "FAVE",
            "className": "fave-button",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "FAVE",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h3>
                Name: 
              </h3>,
              <p>
                Species: 
              </p>,
              <p>
                Homeworld: 
              </p>,
              <p>
                Planet population: 
              </p>,
            ],
            "className": "person card",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Name: ",
                  undefined,
                ],
              },
              "ref": null,
              "rendered": Array [
                "Name: ",
                undefined,
              ],
              "type": "h3",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Species: ",
                  undefined,
                ],
              },
              "ref": null,
              "rendered": Array [
                "Species: ",
                undefined,
              ],
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Homeworld: ",
                  undefined,
                ],
              },
              "ref": null,
              "rendered": Array [
                "Homeworld: ",
                undefined,
              ],
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Planet population: ",
                  undefined,
                ],
              },
              "ref": null,
              "rendered": Array [
                "Planet population: ",
                undefined,
              ],
              "type": "p",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
